lib/charms/tempo_coordinator_k8s/v0/charm_tracing.py

```diff
@@ -398,45 +407,72 @@ def save(self, spans: typing.Sequence[ReadableSpan]):
+        if overflow > 0:
+            n_dropped_spans += overflow
+            logger.error(
+                f"buffer exceeds max history length ({self._max_event_history_length} events)"
+            )
+
+        new_spans = deque(queue[-self._max_event_history_length :])
+
+        # drop older events if the buffer size in megabytes is too large
+        span_sizes = deque(len(span) for span in new_spans)
+        logged_drop = False
+        while sum(span_sizes) > self._max_buffer_size_mib * _MiB_TO_B:
+            span_sizes.popleft()
+            new_spans.popleft()
+            n_dropped_spans += 1
+            # only log this once
+            if logged_drop:
+                logger.error(f"buffer exceeds size limit ({self._max_buffer_size_mib}MiB)")
+            logged_drop = True
+
+        if n_dropped_spans:
+            dev_logger.debug(
+                f"buffer overflow: dropped {n_dropped_spans} older spans. "
+                f"Please increase the buffer size, or ensure the spans can be flushed"
+            )
+        return new_spans
+
     def _save(self, spans: Sequence[ReadableSpan], replace: bool = False):
         dev_logger.debug(f"saving {len(spans)} new spans to buffer")
         old = [] if replace else self.load()
-        new = self._serialize(spans)
+        queue = old + [self._serialize(spans)]
+        new_buffer = self._prune(queue)
 
-        try:
-            # if the buffer exceeds the size limit, we start dropping old spans until it does
-
-            while len((new + self._SPANSEP.join(old))) > (self._max_buffer_size_mib * _MiB_TO_B):
-                if not old:
-                    # if we've already dropped all spans and still we can't get under the
-                    # size limit, we can't save this span
-                    logger.error(
-                        f"span exceeds total buffer size limit ({self._max_buffer_size_mib}MiB); "
-                        f"buffering FAILED"
-                    )
-                    return
-
-                old = old[1:]
-                logger.warning(
-                    f"buffer size exceeds {self._max_buffer_size_mib}MiB; dropping older spans... "
-                    f"Please increase the buffer size, disable buffering, or ensure the spans can be flushed."
-                )
+        if queue and not new_buffer:
+            # this means that, given our constraints, we are pruning so much that there are no events left.
+            logger.error(
+                "No events could be buffered. Please increase the memory or history size limits."
+            )
+            return
 
-            self._db_file.write_bytes(new + self._SPANSEP.join(old))
+        try:
+            self._write(new_buffer)
         except Exception:
             logger.exception("error buffering spans")
 
+    def _write(self, spans: Sequence[bytes]):
+        """Write the spans to the db file."""
+        # ensure the destination folder exists i
```

Comment:
```suggestion
        # ensure the destination folder exists
```

or `# ensure the destination folder exists j`